### PR TITLE
[InstCombine] fold `(a == c && b != c) || (a != c && b == c))` to `(a == c) == (b != c)`

### LINKED ISSUE #92966: Missed optimization for `(a == 0 && b != 0) || (a != 0 && b == 0))`
Condition like `(a == 0 && b != 0) || (a != 0 && b == 0))` could be folded into `(a == 0) != (b == 0)`, but llvm misses this optimization. 

For following function
```c
int test2(int a, int b)
{
    if ((a == 0 && b != 0) || (a != 0 && b == 0))
        return 1;

    return 0;
}
```

llvm generates

```llvm
define dso_local range(i32 0, 2) i32 @src(i32 noundef %0, i32 noundef %1) local_unnamed_addr #0 {
  %3 = icmp eq i32 %0, 0
  %4 = icmp ne i32 %1, 0
  %5 = and i1 %3, %4
  %6 = icmp ne i32 %0, 0
  %7 = icmp eq i32 %1, 0
  %8 = and i1 %6, %7
  %9 = or i1 %5, %8
  %10 = zext i1 %9 to i32
  ret i32 %10
}
``` 

While there exists more optimal IR for such function: 

```llvm
define dso_local range(i32 0, 2) i32 @dst(i32 noundef %0, i32 noundef %1) local_unnamed_addr #0 {
  %3 = icmp eq i32 %0, 0
  %4 = icmp eq i32 %1, 0
  %5 = xor i1 %3, %4
  %6 = zext i1 %5 to i32
  ret i32 %6
}
```

Alive shows that such transformation is correct: https://alive2.llvm.org/ce/z/5AUAao

Odd enough that clang performs such folding if `0` is not a constant, but a function argument: https://godbolt.org/z/PsWh1rP8d

### PR DESCRIPTION
**Pull request by zjaffal**
resolves https://github.com/llvm/llvm-project/issues/92966

alive proof
https://alive2.llvm.org/ce/z/bLAQBS

### CODE REVIEW

**Comment by dtcxzyw** [comment]
@zjaffal @pskrgag 
Does this pattern exist in some real-world applications? Can we generalize it into `(A & B) | (!A & !B) -> A == B`?

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3434]: `Pred` is an out parameter.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3444]: Is this not a generalization of essential `(and/or Same, Same)`?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3444]: > Is this not a ~~generalization~~ specialization of essential `(and/or Same, Same)`?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3444]: Oh I see, you swap preds.

**Comment by pskrgag** [comment]
Hello, @dtcxzyw!

I noticed that pattern during core review on my job and was curios if it can be optimized. IIRC the check was about `buffer == NULL` and `size == 0`.

I believe, it can be generalized.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (pskrgag) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3437]: Shouldn't this fold also work for any other value? 

I am not really familiar with llvm internals, but looks like code is specific to 0 constant

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3437]: Yeah, I think the current implementation is too specific. It also works for pointer compares `(a == nullptr && b != nullptr) || (a != nullptr && b == nullptr)`.

We can reuse the `isKnownInversion` helper here.

Comment (zjaffal) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3437]: I can't find `isKnownInversion` any where being used do you have examples for it? 

I  tried checking on alive to see a compare against any value and it works. 
https://alive2.llvm.org/ce/z/C7Tann

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3437]: See https://github.com/llvm/llvm-project/pull/93591

**Comment by zjaffal** [comment]
I used isKnownInversion to check for both operands and added tests to check against a constant and to check for pointer compare

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3437]: Declare variables where you initialize them.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3447]: You shouldn't assume that the instructions are comparisons here. If you use isKnownInversion, treat them as opaque values. You want to produce `icmp eq Cmp1, Cmp2` here.

File: llvm/test/Transforms/InstCombine/fold-a-or-b-zero.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/fold-a-or-b-zero.ll:22]: Only use `@use()` for multiple-use tests. Simply return the value in normal tests.

**Comment by pskrgag** [comment]
Also there is a small typo in pr title: `a == 1 && b != 0` -> `a == 0 && b != 0`.

**Comment by nikic** [comment]
Please also add an alive2 proof to the PR description.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (zjaffal) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3447]: shouldn't it be`icmp eq cmp1, inv(cmp2)`  ?

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3447]: No, see https://alive2.llvm.org/ce/z/JquFgq.

Comment (zjaffal) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3447]: yeah i get it

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (zjaffal) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3437]: i guess those aren't needed now

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3430]:
```suggestion
  if (!match(Op0, m_OneUse(m_And(m_Value(Cmp1), m_Value(Cmp2)))) ||
      !match(Op1, m_OneUse(m_And(m_Value(Cmp3), m_Value(Cmp4)))))
```

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3435]:
```suggestion
  if (!(isKnownInversion(Cmp1, Cmp3) && isKnownInversion(Cmp2, Cmp4)) ||
      !(isKnownInversion(Cmp1, Cmp4) && isKnownInversion(Cmp2, Cmp3)))
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3425]: Since this is only for `or` at the moment, think an assert of `I.getOpcode` would be useful to avoid future misreads.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3471]: This comment no longer captures the generalization of the impl.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3424]: Static.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3434]:
```suggestion
  if (!match(I.getOperand(0), m_And(m_Value(Cmp1), m_Value(Cmp2))) ||
      !match(I.getOperand(1), m_And(m_Value(Cmp3), m_Value(Cmp4))))
```
No real need to introduce Op0 and Op1 variables to use them in a single place...

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3437]:
```suggestion
  // Check if any two pairs of the and operations are invertions of each other.
```

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3447]:
```suggestion
  if (isKnownInversion(Cmp1, Cmp3) && isKnownInversion(Cmp2, Cmp4))
    return Builder.CreateXor(Cmp1, Cmp4);
  if (isKnownInversion(Cmp1, Cmp4) && isKnownInversion(Cmp2, Cmp3))
    return Builder.CreateXor(Cmp1, Cmp3);
  return nullptr;
```
Would be simpler like this I think...

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3479]:
```suggestion
  // (A == c & B != d) | (A != c & B == d)) -> (A == c) ^ (B == d)
```

**Review by nikic — COMMENTED** [review]
Looks mostly good, a few more nits.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3472]: very nit: can you be consistent on capitolization?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3472]: Also, I guess can you generalize the comment, its really '(A & B) | (C & D) where A == ~C && B == ~D or A == ~D and B == ~D'

Finally, I think you can remove: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp#L1843 w/ this patch.

Comment (zjaffal) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3472]: removing it made the following test fail 
```
define i32 @and_to_xor4(i32 %a, i32 %b) {
; CHECK-LABEL: @and_to_xor4(
; CHECK-NEXT:    [[AND2:%.*]] = xor i32 [[B:%.*]], [[A:%.*]]
; CHECK-NEXT:    ret i32 [[AND2]]
;
  %or = or i32 %b, %a
  %and = and i32 %a, %b
  %not = xor i32 %and, -1
  %and2 = and i32 %not, %or
  ret i32 %and2
}

```

probably I need to move where I call my function to be called there as well so I make sure any negations are pushed inside

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3472]: What is the new result? Is it just `xor i32 %a, %b`? Or a regression?

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3472]: But yeah, to avoid any regressions you could probably just place the function call where the old impl was.

Comment (zjaffal) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3472]: to be able to do that we need to generalise `isKnownInversion` to handle the case of a  and ~a being inversions of each other. I will change that once we land this

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3472]: Oh wow, had just assumed `isKnownInversion` handled that case.

Personally would just do it all in this PR, but up to you.

The code as is now LGTM.

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3472]: > Oh wow, had just assumed `isKnownInversion` handled that case.
> 
> Personally would just do it all in this PR, but up to you.
> 
> The code as is now LGTM.

You told me not to handle this pattern :)
See https://github.com/llvm/llvm-project/pull/93591#issuecomment-2137997343

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3472]: I lack foresight :)

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3424]: What does `Const` mean here?

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!
Please wait for additional approval from other reviewers :)

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3472]: Separate PR please.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3434]:
```suggestion
  // Check if any two pairs of the and operations are inversions of each other.
```

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3424]:
```suggestion
static Value *foldOrOfInversions(BinaryOperator &I,
```
Maybe?

**Review by nikic — APPROVED** [review]
LGTM

Note that there is a conjugate pattern with or and and swapped: https://alive2.llvm.org/ce/z/mRhkK2 Would be good to handle it in a followup as well.

**Pull request merged**
